src/main/java/com/mraof/minestuck/block/redstone/RotatorBlock.java

```diff
@@ -0,0 +1,99 @@
+ * Whether it rotates clockwise or counterclockwise depends on the ROTATION_FLIPPED property which can be cycled by right clicking
+ */
+public class RotatorBlock extends MSDirectionalBlock
+{
+	public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
+	public static final BooleanProperty ROTATION_FLIPPED = BlockStateProperties.INVERTED;
+	
+	public RotatorBlock(Properties properties)
+	{
+		super(properties);
+		this.registerDefaultState(stateDefinition.any().setValue(FACING, Direction.UP).setValue(POWERED, false).setValue(ROTATION_FLIPPED, false));
+	}
+	
+	@Override
+	public void neighborChanged(BlockState state, World worldIn, BlockPos pos, Block blockIn, BlockPos fromPos, boolean isMoving)
+	{
+		super.neighborChanged(state, worldIn, pos, blockIn, fromPos, isMoving);
+		if(fromPos != pos.relative(state.getValue(FACING)))
+			updatePower(worldIn, pos);
+	}
+	
+	@Override
+	public void onPlace(BlockState state, World worldIn, BlockPos pos, BlockState oldState, boolean isMoving)
+	{
+		super.onPlace(state, worldIn, pos, oldState, isMoving);
+		updatePower(worldIn, pos);
+	}
+	
+	@Override
+	@SuppressWarnings("deprecation")
+	public ActionResultType use(BlockState state, World worldIn, BlockPos pos, PlayerEntity player, Hand hand, BlockRayTraceResult hit)
+	{
+		if(!player.isCrouching())
+		{
+			worldIn.setBlock(pos, state.cycle(ROTATION_FLIPPED), Constants.BlockFlags.NOTIFY_NEIGHBORS);
+			if(state.getValue(ROTATION_FLIPPED))
+				worldIn.playSound(null, pos, SoundEvents.PISTON_EXTEND, SoundCategory.BLOCKS, 0.5F, 1.2F);
+			else
+				worldIn.playSound(null, pos, SoundEvents.PISTON_CONTRACT, SoundCategory.BLOCKS, 0.5F, 1.2F);
+			return ActionResultType.SUCCESS;
+		}
+		
+		return ActionResultType.PASS;
+	}
+	
+	public void updatePower(World worldIn, BlockPos pos)
+	{
+		if(!worldIn.isClientSide)
+		{
+			BlockState state = worldIn.getBlockState(pos);
+			boolean hasPower = worldIn.hasNeighborSignal(pos);
+			
+			boolean isPoweredBeforeUpdate = state.getValue(POWERED);
+			
+			if(state.getValue(POWERED) != hasPower)
+				worldIn.setBlockAndUpdate(pos, state.setValue(POWERED, hasPower));
+			else worldIn.sendBlockUpdated(pos, state, state, 2);
+			
+			if(!isPoweredBeforeUpdate && hasPower)
+			{
+				BlockPos facingPos = pos.relative(state.getValue(FACING));
+				BlockState facingState = worldIn.getBlockState(facingPos);
+				BlockState rotatedFacingState;
+				if(state.getValue(ROTATION_FLIPPED))
+					rotatedFacingState = facingState.rotate(worldIn, facingPos, Rotation.COUNTERCLOCKWISE_90);
+				else
+					rotatedFacingState = facingState.rotate(worldIn, facingPos, Rotation.CLOCKWISE_90);
+				
+				if((rotatedFacingState.canSurvive(worldIn, pos) && !rotatedFacingState.hasLargeCollisionShape() && (PistonBlock.isPushable(rotatedFacingState, worldIn, facingPos, null, false, null)) ||
+						(rotatedFacingState.getBlock() instanceof MSDirectionalBlock || rotatedFacingState.getBlock() instanceof LogicGateBlock)))
```

Comment:
Is this special treatment really needed? (Referring to the instanceof checks with block classes) Can you express this in conditions like those on the line before, so as to not hardcore it? Or if you can't get away from giving special treatment to certain kinds of blocks, could you create a block tag for it?

Reply:
A block tag makes sense, I wanted to place those sort of limitations in order to minimize potential compatibility issues